README.md
dejavu
------

[![Build Status](https://secure.travis-ci.org/IndigoUnited/dejavu.png)](http://travis-ci.org/IndigoUnited/dejavu)

Have you ever had the feeling that you're seeing something you've already seen
before? That's the feeling you get when using `dejavu`.

If you are a developer coming from a language like PHP, Java, ActionScript 3.0,
and others, it's likely that you are already familiar with Object Oriented
Programming. However, JavaScript uses prototypal inheritance which, although
powerful and flexible, can be difficult to unde
rstand, and specially to maintain
in large projects.

`dejavu` is a library that delivers classical inheritance on top of JavaScript
prototypal inheritance, making it a breeze to move into JavaScript.



## Why another?

There are some libraries out there able to shim classical inheritance,
however none offers all the functionality that many programmers require.

Also, even though being one of the most feature rich OOP libraries out there, it has an outstanding performance, rivaling with vanilla in production.



## Features

* Classes (concrete, abstract and final)
* Interfaces
* Mixins (so you can get some sort of multiple inheritance)
* Private and protected members
* Static members
* Constants
* Context binding for functions
* Method signature checks
* Possible to extend or borrow from vanilla classes
* Custom instanceOf with support for Interfaces
* Two builds, `regular` and `AMD` based
    * `AMD` optimized for speeding up developer workflow, allowing testing
      without the need to re-compile everything into a single file
    * `regular` if you are not using `AMD` in your projects
* Two modes for each build, `strict` and `loose`
    * `strict` best in development, enforcing a lot of checks, making sure you
      don't make many typical mistakes
    * `loose` best for production, without checks, improving performance

Users are encouraged to declare
['use strict'](https://developer.mozilla.org/en/JavaScript/Strict_mode) while
using the `dejavu` in strict mode, otherwise some code might fail silently.
This can happen because `dejavu` uses `Object.freeze` and `Object.seal` to lock
classes and instances, guaranteeing that no one changes the behaviour of your
classes by replacing methods, etc, and possibly breaking your code, making it
really hard to pin point what's wrong. Although this is the default behaviour, it can be changed.

You will read more on it later in this document.

**Do not confuse 'use strict' with the dejavu strict mode.**



## Getting started

The quickest way to start using `dejavu` in your project, is by simply including
`dist/regular/strict/dejavu.js` (note that this is in __strict__ mode).

If you're developing a __client-side__ app, simply put the file in some folder,
and include it in the HTML:

```html
<!DOCTYPE html>
<html>
    <head>
        <meta http-equiv="Content-Type"
              content="text/html; charset=UTF-8">
        <script type="text/javascript" src="dejavu.js"></script>
    </head>
    <body>
        <script type="text/javascript">
            'use strict';

            // declare the "Person" class
            var Person = dejavu.Class.declare({
                _name: null,

                initialize: function(name) {
                    this.setName(name);
                },

                setName: function(name) {
                    this._name = name;

                    return this;
                },

                getName: function() {
                    return this._name;
                }
            });

            // create a new instance of person
            var indigo = new Person('Marco');
            console.log('A new indigo was born,', indigo.getName());
        </script>
    </body>
</html>
```

This will make a `dejavu` global available for you.
If you're developing in __Node.js__, install it with `npm install dejavu` and use it like so:

```js
var dejavu = require('dejavu');

// declare the "Person" class
var Person = dejavu.Class.declare({
    _name: null,

    initialize: function(name) {
        this.setName(name);
    },

    setName: function(name) {
        this._name = name;

        return this;
    },

    getName: function() {
        return this._name;
    }
});

// create a new instance of person
var indigo = new Person("Marco");
console.log("A new indigo was born,", indigo.getName());
```

The installation will create a `.dejavurc` in your package, where you can enable/disable the strict mode as well as change other dejavu options.
By default the strict mode is used. Still, you want to leverage your package with the loose mode.
Because dejavu reads`.dejavurc` from the process.cwd(), packages that require your package will be running the loose mode unless they also define a `.dejavurc` (which will only happen if they also depend on dejavu directly).



## Performance

Since all those nice features and common rules of classic OOP degrade
performance, `dejavu` has two separates modes, for different stages in the
development.

The `strict` mode is suitable for __development__, and will do all sorts of
checks, throwing an error when you try to do something considered illegal.

**Note that if your project works in strict mode, it will work in loose mode.**

As for the `loose` mode, there is no overhead associated with checks, thus
making it suitable for __production__, since it will be more efficient and
have a __lower memory footprint and filesize__.

Finally, in order to achieve that extra edge, that puts `dejavu` next to vanilla
JS in terms of performance, you should run the optimizer that is bundled with
the library. Note that this step is completely optional, and `dejavu` will still
perform faster than most libraries, even if you don't run the [optimizer](https://github.com/IndigoUnited/dejavu-optimizer).
The optimizer will analyse your code and make some improvements,
boosting it a bit further.

You can check the benchmarks in [jsperf](http://jsperf.com/oop-benchmark/107)
comparing `dejavu` with other OOP libraries. Note that the loose mode
is used in this test, simulating a production environment, and both the normal
and optimized versions are tested. It is also important to mention that many libraries like
[JSFace](https://github.com/tnhu/jsface) does not chain prototypes. This gives
JSFace an extra edge in performance in some browsers, like Firefox, but renders
the `instanceof` operator useless.



## Syntax

### Overview

Here's an overview of what most developers look for in an OOP library. You can find complete examples further down.

```js
var Person = Class.declare({
    // although not mandatory, it's really useful to identify
    // the class name, which simplifies debugging
    $name: 'Person',

    // this is a protected property, which is identified by
    // the single underscore. two underscores denotes a
    // private property, and no underscore stands for public
    _name: null,
    __pinCode: null,

    // class constructor
    initialize: function (name, pinCode) {
        this._name = name;
        this.__pinCode = pinCode;

        // note that we're binding to the current instance with `$bind`.
        // this will be explained with great detail later on
        setTimeout(this._logName.$bind(this), 1000);
    },

    // public method (follows the same visibility logic, in this case with no underscore)
    getName: function () {
        return this._name;
    }

    _logName: function () {
        console.log(this._name);
    }
});
```

### Complete example

For those looking for something more, here's a more complete usage of `dejavu`.

This example illustrates the usage of:

- `$name` meta attribute
- `this.$self` vs `this.$static`
- `$super()` for accessing overridden methods
- `instanceOf`
- member visibility
- statics, abstracts, abstract statics, finals, final statics and constants
- `$extends` vs `$borrows`

In this case, and keep in mind that this is just for illustration purposes, we'll create three interfaces, that are implemented by an abstract class, that is then extended by a concrete class.


```js
var dejavu = require('dejavu');

// ------------ AN INTERFACE ------------
// this interface is useless, is only here to illustrate
// that interfaces can extend other interfaces
var UselessInterface = dejavu.Interface.declare({
    $name: 'UselessInterface'
});

var PersonInterface = dejavu.Interface.declare({
    $name: 'PersonInterface',
    // if you need to extend multiple interfaces,
    // just provide an array
    $extends: UselessInterface,

    // interface methods can specify argument list, and any class
    // that implements that interface will be automatically checked,
    // to make sure it obeys the method signature. If you want to
    // specify an optional argument, you should prepend it by a dollar
    // sign, like so: someMethod(arg1, arg2, $thisArgIsOptional)
    getName: function () {},
    setName: function (name) {}
});

// ------------ ANOTHER INTERFACE ------------
var EngineerInterface = dejavu.Interface.declare({
    $name: 'EngineerInterface',

    think: function(subject) {}
});

// ------------ AN ABSTRACT CLASS ------------
var AbstractIndigo = dejavu.AbstractClass.declare({
    $name: 'AbstractIndigo',
    // implements multiple interfaces
    $implements: [PersonInterface, EngineerInterface],

    $constants: {
        INDIGO_WEBSITE: 'http://www.indigounited.com/',
        INDIGO_EMAIL:   'hello@indigounited.com'
    },

    $statics: {
        logIndigoInfo: function () {
            // by using this.$static, we're making sure that dejavu
            // uses late binding to resolve the member. If you're
            // looking for early binding, you can use this.$self
            // instead
            console.log(
                this.$static.INDIGO_WEBSITE,
                this.$static.INDIGO_EMAIL
            );
        }
    },

    // method/attribute visibility is controlled by
    // the number of underscores that the identifier
    // has:
    // public:    no underscores
    // protected: 1 underscore
    // private:   2 underscores
    //
    // the attribute below is protected
    _name: null,

    getName: function () {
        return this._name;
    },

    setName: function (name) {
        this._name = name;

        return this;
    },

    // note that we're not implementing the method `think()` of the
    // EngineerInterface. This will be automatically turned into an
    // abstract method, since we're in an abstract class
    $abstracts: {
        beAwesome: function () {}

        // you can also put "$statics {}" here
        // to create an abstract static method
    },

    // finals are not overridable
    $finals: {
        // you can also put "$statics {}" here
        // to create a final static method

        thisIsFinal: function () {
            console.log('Can\'t change this!');
        }
    }
});

// ------------ A CONCRETE CLASS ------------
// also, if you need this concrete class to be final,
// you can just use dejavu.FinalClass.declare instead
var Indigo = dejavu.Class.declare({
    $name: 'Indigo',
    // class extends another one.
    //
    // in case you need to extend from several classes,
    // you can instead use $borrows, and specify an
    // array of identifiers. Still, note that borrowing
    // will not allow you to perform dejavu.instanceOf
    // tests, as the class is not technically extending
    // the other, just borrowing its behaviour.
    $extends: AbstractIndigo,

    _subject: 'nothing',

    initialize: function (name) {
        // call the parent method, in this case the parent constructor,
        // but can be applied to any method when you need to call the
        // overridden method
        this.$super();

        this.setName(name);

        this._logThought();
    },

    beAwesome: function () {
        console.log(this._name, 'is being awesome!');
        this.$self.logIndigoInfo();
        this.think('the next big thing');
    },

    think: function (subject) {
        this._subject = subject;
    },

    _logThought: function () {
        console.log(this._name, 'is thinking about', this._subject);
    }
});

var indigo = new Indigo('André');
indigo.beAwesome();

// check the type of an object
console.log(
    dejavu.instanceOf(indigo, EngineerInterface) ?
    'we have an engineer!'
    : 'say what now?'
);
console.log(dejavu.instanceOf(indigo, Indigo) ?
    'we have an indigo!'
    : 'say what now?'
);
// native instanceof also works for classes, but not for interfaces
console.log((indigo instanceof Indigo) ?
    'we have an indigo!'
    : 'say what now?'
);

```



## Taking it to another level

Front-end devs are encouraged to program using the AMD paradigm because of its obvious benefits.
Since dejavu is built on it, it will integrate seamlessly with your AMD loader.

```js
{
    // your loader config
    paths: {
           'mout': '../vendor/mout/src'
    },
    packages: [
        {
            name: 'dejavu',
            // You can switch to the loose mode anytime
            location: '/path/to/dejavu/dist/strict'
        }
    ]
}
```

With this setup, you can require the dejavu object or specific parts of it:

```js
// Load dejavu completely
define(['dejavu'], function (dejavu) {

    // the dejavu variable is an object that contains:
    // Class
    // FinalClass
    // AbstractClass
    // Interface
    // instanceOf
    // options

    // example usage
    var Person = dejavu.Class.declare({
        initialize: function () {
            // ...
        }
    });

    return Person;
});

// In this case, only the `Class` module of `dejavu` is included,
// which means all the other modules are not loaded.
define(['dejavu/Class'], function (Class) {

    // Example usage
    var Person = Class.declare({
        initialize: function () {
            // ...
        }
    });

    return MyClass;
});
```



## Additional details

### Binding and anonymous members

You will eventually run into a situation where you want to declare a callback that accesses class members. On traditional JavaScript, you would just `var that = this` or `.bind(this)`, and everything would be ok, because there is no restriction on visibility. Since `dejavu` enforces this, you will need to mark that callback as a member of the class, using something like the following:

```js
// ...
var that = this;
setTimeout(function () {
    that._someProperty = 'protected properties on callbacks';
}.$member(), 1000);
```

```js
// ...
setTimeout(function () {
    this._someProperty = 'protected properties on callbacks';
}.$member().bind(this), 1000);
```

If the `$member().bind(this)` is too verbose for you, you can just `.$bind(this)`, which is equivalent.

Finally, when defining a method directly on the class declaration that you know will always be used using the class context, you can bind it right there like so:

```js
var MyClass = dejavu.Class.declare({
    $name: 'MyClass',

    doSomething: function () {
        // notice that _someMethod is $bound() below,
        // which is more efficient than $bind()ing on
        // every execution of doSomething()
        setTimeout(this._someMethod, 1000);
    },

    _someMethod: function () {
        console.log('method efficiently bound');
    }.$bound()
});
```

### Classes/instances are locked

By default, constructors and instances are locked. This means that no one can monkey patch your code.

This behaviour can be changed in two ways:

#### With the $locked flag:

```js
var UnlockedIndigo = Class.declare({
    $name: 'UnlockedIndigo',
    $locked: false

    initialize: function () {
        // Altough the foo property is not declared,
        // it will not throw an error

        this.name = 'Filipe';
    },

    talk: function () {
        console.log('An indigo is talking!');
    }
});
```

Members can be added, replaced and deleted from the prototype:

```js
UnlockedIndigo.prototype.age = 20;
UnlockedIndigo.prototype.talk = function () {
    console.log('... now is running');
};
```

Members can be added, replaced and deleted from the instance:

```js
var filipe     = new UnlockedIndigo();
filipe.friends = ['Marco','Andre'];
filipe.talk    = function () {
    console.log('I am talking about dejavu!');
};
```

#### By setting the global option:

This will change the default behaviour, but classes can still override it with the $locked flag.

```js
dejavu.options.locked = false;
```

Note that once a class is unlocked, its subclasses cannot be locked.
Also, although undeclared members are allowed, they will not have their access controlled (they are interpreted as public).



### Vanilla classes

`dejavu` allows you to extend or borrow vanilla classes. In this case, constructors and instances are forcibly UNLOCKED.

```js
function Person(name) {
    this.name = name;
};

var Engineer = dejavu.Class.declare({
    $extends: Person
});

var filipe = new Engineer('Filipe');
// Engineer class and filipe instance are unlocked

```


### Debugging

In strict mode, logging instances with `console.log` will print some strange stuff (getters, setters, etc).
This happens because `dejavu` manages accesses to private/protected members as well as make other stuff work.
To get around this issue, `dejavu` provides a `console.inspect` method that works just like `console.log` but prints a friendly object representation of the instance.



## Works on

* IE (6+)
* Chrome (4+)
* Safari (3+)
* Firefox (3.6+)
* Opera (9+)
* Node.js and Rhino



## Dependencies

dejavu depends on [mout](https://github.com/mout/mout).
If you use the regular build, you don't need to worry because all functions used from mout are bundled for you.
If you use the AMD build, learn [how](https://github.com/IndigoUnited/dejavu#taking-it-to-another-level) to setup your loader.
If you use dejavu on `node`, `npm` will take care of fetching everything for you.


## Building dejavu

Simply run `npm install` to install all the tools needed.
Then just run `npm run-script build`.



## Testing dejavu

`dejavu` has more than 250 tests.
Simply run `npm install` to install all the tools needed.
Then just run `npm test` to execute them.



## License

Released under the [MIT License](http://www.opensource.org/licenses/mit-license.php).


CHANGELOG.md
# Changelog

### 0.4.2 - 2013-05-09
- Change from `component.json` to `bower.json`.

###0.4.1 - 2013-04-09
- Fix `instanceOf` throwing an exception when passing non-constructors.

### 0.4.0 - 2013-03-12
- Move `optimizer` to its own [repository](https://github.com/IndigoUnited/dejavu-optimizer)

### 0.3.9 - 2013-03-10
- Add minified version of the `regular loose` mode.
- Fix some error message typos.

### 0.3.8 - 2013-03-09
- Fix `console.inspect` not handling circular references on arrays/objects

### 0.3.7 - 2013-03-07
- Fix `.dejavurc` strict value not being correctly interpreted in some edge cases (only affects `node` users)
- Fix an important bug when a class extends another one of a different `node` package in `strict` mode (only affects `node` users)
- Refactor tests and make them run in all environments

### 0.3.6 - 2013-03-06
- Remove micro-optimization in non-optimized classes in loose version (it was giving some problems)

### 0.3.5 - 2013-03-05
- Add `grunt` task for the optimizer
- Small improvements to the optimizer

###0.3.4 - 2013-03-04
- Fix optimizer error when using `this.$super.apply` or `.call`

###0.3.3 - 2013-02-18
- Fix `postinstall` script in some rare cases

### 0.3.2 - 2013-02-15
- Fix rare bug in when using `$bind` with `null` values

### 0.3.1 - 2013-02-02
- Fix bug in the optimizer, not guessing when to use the non-closure when correctly
- Other minor optimizer tweaks

### 0.3.0 - 2013-02-02
- __Add $member() to allow common usage of `var that = this` and still access protected/private members__
- Fix dejavu being unlocked by default in `node`
- Loose mode now keeps `$name`
- Other minor tweaks and improved documentation

### 0.2.2 - 2013-01-22
- Fix one more issue with the optimizer task in windows

### 0.2.1 - 2013-01-19
- Fix optimizer automaton task in windows

js/main.js
/*jshint regexp:false*/
/*global Documentation, Browserscope*/

// Website version
if (!window.siteVersion) {
    window.siteVersion = +(new Date());
}

// Documentation parsing
(function () {
    var leftColumnEl,
        rightColumnEl,
        isPhone,
        hash;

    isPhone = (function () {
        var uAgent = navigator.userAgent.toLowerCase(),
            isMobile = false;

        if (/android|tablet|avantgo|bada\/|blackberry|blazer|compal|elaine|fennec|hiptop|iemobile|ip(hone|od|ad)|iris|kindle|lge |maemo|meego.+mobile|midp|mmp|netfront|opera m(ob|in)i|palm( os)?|phone|p(ixi|re)\/|plucker|pocket|psp|series(4|6)0|symbian|treo|up\.(browser|link)|vodafone|wap|windows (ce|phone)|xda|xiino|dell streak|playbook|silk/.test(uAgent)) {
            isMobile = true;
        } else if (/1207|6310|6590|3gso|4thp|50[1-6]i|770s|802s|a wa|abac|ac(er|oo|s\-)|ai(ko|rn)|al(av|ca|co)|amoi|an(ex|ny|yw)|aptu|ar(ch|go)|as(te|us)|attw|au(di|\-m|r |s )|avan|be(ck|ll|nq)|bi(lb|rd)|bl(ac|az)|br(e|v)w|bumb|bw\-(n|u)|c55\/|capi|ccwa|cdm\-|cell|chtm|cldc|cmd\-|co(mp|nd)|craw|da(it|ll|ng)|dbte|dc\-s|devi|dica|dmob|do(c|p)o|ds(12|\-d)|el(49|ai)|em(l2|ul)|er(ic|k0)|esl8|ez([4-7]0|os|wa|ze)|fetc|fly(\-|_)|g1 u|g560|gene|gf\-5|g\-mo|go(\.w|od)|gr(ad|un)|haie|hcit|hd\-(m|p|t)|hei\-|hi(pt|ta)|hp( i|ip)|hs\-c|ht(c(\-| |_|a|g|p|s|t)|tp)|hu(aw|tc)|i\-(20|go|ma)|i230|iac( |\-|\/)|ibro|idea|ig01|ikom|im1k|inno|ipaq|iris|ja(t|v)a|jbro|jemu|jigs|kddi|keji|kgt( |\/)|klon|kpt |kwc\-|kyo(c|k)|le(no|xi)|lg( g|\/(k|l|u)|50|54|\-[a-w])|libw|lynx|m1\-w|m3ga|m50\/|ma(te|ui|xo)|mc(01|21|ca)|m\-cr|me(di|rc|ri)|mi(o8|oa|ts)|mmef|mo(01|02|bi|de|do|t(\-| |o|v)|zz)|mt(50|p1|v )|mwbp|mywa|n10[0-2]|n20[2-3]|n30(0|2)|n50(0|2|5)|n7(0(0|1)|10)|ne((c|m)\-|on|tf|wf|wg|wt)|nok(6|i)|nzph|o2im|op(ti|wv)|oran|owg1|p800|pan(a|d|t)|pdxg|pg(13|\-([1-8]|c))|phil|pire|pl(ay|uc)|pn\-2|po(ck|rt|se)|prox|psio|pt\-g|qa\-a|qc(07|12|21|32|60|\-[2-7]|i\-)|qtek|r380|r600|raks|rim9|ro(ve|zo)|s55\/|sa(ge|ma|mm|ms|ny|va)|sc(01|h\-|oo|p\-)|sdk\/|se(c(\-|0|1)|47|mc|nd|ri)|sgh\-|shar|sie(\-|m)|sk\-0|sl(45|id)|sm(al|ar|b3|it|t5)|so(ft|ny)|sp(01|h\-|v\-|v )|sy(01|mb)|t2(18|50)|t6(00|10|18)|ta(gt|lk)|tcl\-|tdg\-|tel(i|m)|tim\-|t\-mo|to(pl|sh)|ts(70|m\-|m3|m5)|tx\-9|up(\.b|g1|si)|utst|v400|v750|veri|vi(rg|te)|vk(40|5[0-3]|\-v)|vm40|voda|vulc|vx(52|53|60|61|70|80|81|83|85|98)|w3c(\-| )|webc|whit|wi(g |nc|nw)|wmlb|wonu|x700|yas\-|your|zeto|zte\-/.test(uAgent.substr(0, 4))) {
            isMobile = true;
        }

        if (!isMobile) {
            return false;
        }

        // Exclude if is tablet
        if (/ipad|android 3|dell streak|sch-i800|sgh-t849|gt-p1010|playbook|tablet|silk|kindle fire/.test(uAgent)) {
            return false;
        }

        return true;
    }());

    // Reach to hash chages
    function reactToHash() {
        var newHash = location.hash.replace(/#(section\-)?/, '');
        if (hash === newHash) {
            return;
        }

        hash = newHash;
        if (hash) {
            $.smoothScroll({ scrollTarget: '#' + newHash });
        }
    }

    function parseBlock(els) {
        var blockEl = $('<div class="block"></div>');
        blockEl.append(els);

        return blockEl;
    }

    function getBlockTitle(els) {
        var first = els.get(0),
            tag = first.tagName.toLowerCase();

        if (tag === 'h1' || tag === 'h2') {
            return first.innerHTML;
        }

        return null;
    }

    function addBlock(els) {
        els = $(els);

        var title = getBlockTitle(els),
            headerEl = els.eq(0),
            el,
            slug = title.replace(/\s+/g, '-').replace(/\?/g, '').toLowerCase(),
            aEl = $('<a href="#' + slug + '" id="' + slug + '"></a>').append($('<span></span>'));

        els = els.not(headerEl);
        aEl.children().empty();
        headerEl = aEl.children().eq(0);
        Array.prototype.unshift.call(els, aEl.get(0));
        el = parseBlock(els),
        headerEl.addClass('block-header block-header-' + slug);

        aEl.after($('<div class="clearfix"></div>'));

        switch (title) {
        case 'dejavu':
        case 'Features':
        case 'Getting started':
        case 'Syntax':
            leftColumnEl.append(el);
            break;
        case 'Performance':
        case 'Why another?':
            rightColumnEl.append(el);
            break;
        case 'Benchmarks':
            el.find('p').each(function (i, p) {
                p.innerHTML = p.innerHTML
                    .replace(/\{\{graph\}\}/, '<span class="benchmark chart loading"></span>')
                    .replace(/\{\{graph_mobile\}\}/, '<span class="benchmark-mobile chart loading"></span>');
            });
            rightColumnEl.append(el);
            break;
        default:
            if (leftColumnEl.height() <= rightColumnEl.height()) {
                leftColumnEl.append(el);
            } else {
                rightColumnEl.append(el);
            }
        }

        // Finally add smooth scroll (for the headers)
        aEl.smoothScroll({
            afterScroll: function () {
                hash = slug;
                location.hash = '#section-' + slug;
            }
        });
    }

    function parseDoc(str) {
        // If is phone, the right column is actually the left column
        // After, we also remove the right column
        leftColumnEl = $('#content .left'),
        rightColumnEl = isPhone ? leftColumnEl : $('#content .right');

        if (isPhone) {
            $(document.body).addClass('phone');
            $('#content .right').remove();
        }

        var el = $('<div style="display: none"></div>').html(str),
            children,
            length,
            els,
            x,
            tag,
            curr;

        el.find('hr').remove();                // Remove all hr's
        el.find('a').attr('target', '_blank'); // All links should open a new window

        children = el.children(),
        length = children.length,
        els = [];

        for (x = 0; x < length; x += 1) {
            curr = children.get(x);
            tag = curr.tagName.toLowerCase();
            if ((tag === 'h1' || tag === 'h2') && els.length) {
                addBlock(els);
                els = [];
            }

            els.push(curr);
        }

        if (els.length) {
            addBlock(els);
        }

        reactToHash();
        $(window).on('hashchange', reactToHash);
    }

    window.Documentation = {
        parse: parseDoc
    };
}());


// Browserscore object
// Fetches the perf results and draws the graph
(function () {

    var testId = 'agt1YS1wcm9maWxlcnINCxIEVGVzdBjk7qoTDA',
        cb = '_' + parseInt(Math.random() * 1e9, 10),
        // The blacklist below is due to unstable versions of browsers that should not
        // yet be accounted for
        browserBlacklist = [/iPhone 6\.0\.2/]; // /Firefox 17\./ -> Example of an entry

    function browserIsBlacklisted(browser) {
        var i,
            totalBlacklisted = browserBlacklist.length;

        for (i = 0; i < totalBlacklisted; i += 1) {
            if (browserBlacklist[i].test(browser)) {
                return true;
            }
        }

        return false;
    }

    function drawChart(title, results, el) {
        var browserName, browser, testName, test, line, browserResults, data,
            chart = new google.visualization.BarChart(el.get(0)),
            lines = [],
            headerHash = {},
            header = ['Browser'],
            max = 0,
            val;

        // TODO: add ops/sec to the tooltip
        // TODO: put M

        // Generate data for the graph
        for (browserName in results) {
            if (results.hasOwnProperty(browserName)) {
                browser = results[browserName];
                line = [browserName];
                browserResults = browser.results;
                for (testName in browserResults) {
                    if (browserResults.hasOwnProperty(testName)) {
                        if (!headerHash[testName]) {
                            headerHash[testName] = 1;
                            header.push(testName);
                        }
                        test = browserResults[testName];
                        val = parseInt(test.result, 10);
                        if (max < val) {
                            max = val;
                        }
                        line.push(val);
                    }
                }
                lines.push(line);
            }
        }

        data = [header].concat(lines);
        max += 3 * 1e6;
        el.removeClass('loading');
        chart.draw(google.visualization.arrayToDataTable(data), {
            title: null,
            backgroundColor: '#000',
            fontName: 'Source Sans Pro',
            fontSize: 14,
            chartArea: {
                top: 70,
                left: 125,
                right: 0,
                bottom: 0,
                width: el.width(),
                height: el.height() - 140
            },
            tooltip: {
                textStyle: {
                    color: 'black'
                }
            },
            legend: {
                position: 'top',
                alignment: 'start',
                textStyle: {
                    color: 'white'
                },
                pagingTextStyle: {
                    color: 'white'
                },
                scrollArrows: {
                    activeColor: '#8C008C',
                    inactiveColor: '#250025'
                }
            },
            vAxis: {
                baselineColor: '#8C008C',
                textStyle: {
                    color: 'white'
                }
            },
            hAxis: {
                title: 'ops per second (higher is better)',
                baselineColor: '#8C008C',
                viewWindowMode: 'explicit',
                viewWindow: {
                    max: max
                },
                titleTextStyle: {
                    fontSize: 15,
                    color: 'white'
                },
                textStyle: {
                    fontSize: 14,
                    color: 'white'
                },
                gridlines: {
                    color: '#250025'
                }
            }
        });
    }


    window[cb] = function (response) {
        var results = response.results,
            key,
            name,
            version,
            split,
            mobile = ['android', 'ipad', 'iphone', 'webos touchpad'],
            newResults = {};

        // Parse non-mobile browsers
        // Only keep the most recent browser in the results
        for (key in results) {
            split = key.split(' ');
            name = split.slice(0, split.length - 1).join(' ');
            version = split[split.length - 1];
            version = version ? parseFloat(version) : version;

            // Include mobile, but exclude blacklisted browsers
            if ($.inArray(name.toLowerCase(), mobile) !== -1 || browserIsBlacklisted(key)) {
                continue;
            }

            newResults[name + ' ' + version] = results[key];
        }

        // Draw the chart
        drawChart(response.category_name, newResults, $('.benchmark'));

        // Parse mobile browsers
        // Only keep the most recent browser in the results
        newResults = {};
        for (key in results) {
            split = key.split(' ');
            name = split.slice(0, split.length - 1).join(' ');
            version = split[split.length - 1];
            version = version ? parseFloat(version) : version;

            // Include mobile, but exclude blacklisted browsers
            if ($.inArray(name.toLowerCase(), mobile) === -1 || browserIsBlacklisted(key)) {
                continue;
            }

            newResults[name + ' ' + version] = results[key];
        }

        // Draw the chart
        drawChart(response.category_name, newResults, $('.benchmark-mobile'));
    };

    function fetchData() {
        var script = document.createElement('script'),
            first = document.getElementsByTagName('script')[0];

        script.async = 1;
        script.src = '//www.browserscope.org/user/tests/table/' + testId +
            '?v=3&o=json&callback=' + cb;
        first.parentNode.insertBefore(script, first);
    }

    window.Browserscope = {
        update: fetchData
    };
}());


$(document).ready(function () {
    // Download the tmpl
    var promise = $.ajax('tmpl/doc.tmpl?v=' + siteVersion, {
        timeout: 15000
    }),
        contentEl = $('#content');

    promise.complete(function () {
        contentEl.removeClass('loading');
    });
    promise.fail(function () {
        contentEl.find('.left').html('Oops, something went wrong.');
    });
    promise.success(function () {
        // Parse it
        Documentation.parse(promise.responseText);

        // Highlight code
        // Do not highlight in <=IE8 because highlightjs throws errors on it
        if (!$.browser.msie || $.browser.version > 8)  {
            var blocks = $('pre code'),
                length = blocks.length,
                x;

            for (x = 0; x < length; x += 1) {
                hljs.highlightBlock(blocks.get(x));
            }
        }

        // Get perf results from browserscope
        Browserscope.update();
    });
});